com.unity.netcode.gameobjects/Runtime/Messaging/Bind.cs

```diff
@@ -0,0 +1,14 @@
+using System;
+
+namespace Unity.Netcode
+{
+    public class Bind : Attribute
```

Comment:
I remember talking about this in testing context. If we just want to mock different NetworkManagers, we can do that in different ways without crafting our own DI.
I really do not want to implement features out in the wild if we can stick with existing stuff and can avoid putting more stuff onto table.

Do you mind explaining the scenario and the need for this Bind/SystemOwner and testing scenario for us so that we could have that discussion here or somewhere else and hopefully come up with alternatives?

Reply:
This isn't for mocking network managers... this is for making sure that messages that are written for testing purposes aren't able to be sent and received in production environments. It has nothing to do with mocking, it's ensuring test code doesn't leak into production.

SystemOwner, as I explained above, is how we pass NetworkManager back to the message handlers without introducing strong dependencies between MessagingSystem and NetworkManager.

Reply:
I'm really not sure why DI keeps coming up with relation to Bind because it has nothing to do with DI.

Reply:
> this is for making sure that messages that are written for testing purposes aren't able to be sent and received in production environments.

if that's the case then perhaps macros or moving separate asmdef?
we're rolling out it to runtime, shipping it to users but sounds like we don't need to ship `Bind`.

Reply:
for the `SystemOwner`, we already have that dependency whether it's through `object` or an explicit `NetworkManager` (or `IMessageSender`) type. `SystemOwner` is a DI model where you inject `NetworkManager` or something via that struct.
it is fine to know that we have `NetworkManager` dep in `MessagingSystem`, and there is no way around that. I'm not sure how much value we're actually getting by hiding it behind an opaque `object`.

Reply:
There's no injection happening. Bind doesn't do anything other than tell MessagingSystem which messages to pay attention to when it's doing reflection. SystemOwner is passed in to the MessagingSystem constructor.

MessagingSystem does not need a NetworkManager dependency... the messages do, but the system doesn't. Adding hard dependencies like that makes testing extremely difficult as I wouldn't be able to test MessagingSystem without creating a NetworkManager for it. It's also just generally bad practice, violates SOLID principles, and tends to spiral through the codebase once you start doing it and create the "big ball of mud" problem. Keeping things agnostic about other things as much as possible is cleaner and more maintainable code.

Reply:
And more testable.

Reply:
Renaming this to `IgnoreIfSystemOwnerIsNotOfType`. I think that makes it much more clear what it's doing... I think a lot of the controversy here was because of bad naming.

Reply:
> SystemOwner is passed in to the MessagingSystem constructor

you "inject" a specific instance through constructor and pass it onto invididual messages later. whether MessagingSystem uses it itself or not, what you're reinventing here is DI — except you throw away the type and effectively make it C++ `void*`, C# `object`.

if messages require a specific subset funtionality, that should be an interface, let's say `INetworkMessageSender` and `INetworkMessageHandler` — `NetworkManager` could implement those interfaces and you could rely on them by doing proper type-based DI through the MessagingSystem ctor, then pass them onto individual messages later. that way, you could have mock implementations and you wouldn't require full-blown NetworkManager.

the point is, there are more type-based, C# friendly ways of doing this. we can discover other approaches. I do get that we have very little time to get this in but I'd rather leave this home-brew DI skeleton out for now and do it properly later. if we can't leave this out today, let's call it out, make things internal and rework it post-feature-freeze as a tech debt, because it looks pretty much like a tech debt to me.

Reply:
on the `Bind` front, I'm going to try to kill it locally and see what breaks/works. I think we can just not have it but wrap testing-only stuff inside either an asmdef or some macros to avoid it from landing into shipping/runtime.

Reply:
Renamed.

Reply:
I'm not going to kill SystemOwner unless you have a specific alternative to avoid MessagingSystem having a direct dependency on NetworkManager... I'm not going to budge on avoiding hard dependencies here. The only reason SystemOwner even exists is because the message handlers HAVE to be able to reference the correct network manager, but MessagingSystem does not need to know anything about NetworkManager... so SystemOwner is an opaque type.

Making MessagingSystem depend on NetworkManager when he doesn't need it makes testing MessagingSystem 100x harder, means we can't test it without creating a NetworkManager - can't mock it because MessagingSystem would require exactly NetworkManager at that point (or a subclass of). It doesn't have any interface dependencies on NetworkManager, it just passes it blindly to the messages, so any interface to describe for testing is just going to be empty.

Test messages require no functionality. Real messages require basically all of NetworkManager, so describing the interface for them again adds difficulty for testing.

But beyond all of the academic discussion, Matt specifically asked me to not have MessagingSystem depend on NetworkManager in any way.

Reply:
As far as `Bind` goes, wrapping the test messages in macros still doesn't really solve the problem, because the test messages also need to be isolated from each other... if a test is running more code than what it's testing, it's an invalid test. A failure from one test should not cause a failure for another. A macro around all test messages would violate that... `IgnoreMessageIfSystemOwnerNotOfType` achieves true test isolation, which is important... I really don't understand why an attribute that only applies to test messages and has no bearing on the actual runtime behavior of the system is so contentious...

Reply:
Again - would like to continue this conversation after feature freeze since this is all internal types anyway.